typescript/patterns/src/xa_transactions.ts

```diff
@@ -0,0 +1,123 @@
+// thus participate consistently in a sequence of steps in a workflow or durable
+// handler execution. 
+//
+// To do that, we use the 2-phase-commit protocol approach, implemented by many
+// databases via the XA transaction standard. For example, PostgreSQL supports this
+// via 'prepared transactions': https://www.postgresql.org/docs/current/sql-prepare-transaction.html
+//
+// Unlike the classical 2-phase-commit scenario, we are not using the protocol to
+// commit according to vote 
+
+// In principle, the operation is as simple as
+// (1) sideEffect: (exequte query, prepare transaction under ID, return ID from sideEffect).
+//                 This side effect commits once, to a single ID is stored in Restate for
+//                 committing.
+// (2) sideEffect: commit the prepared transaction
+//
+// The slightly higher complexity here comes from the fact that we need to ensure that
+// any transaction that was prepared, but will not be committed, must be cleaned up by
+// the recovery run. Otherwise that transaction will keep holding database locks and
+// block other transactions. We also re-try the transaction after that.
+
+// ----------------------------------------------------------------------------
+
+// The methods here must have the same semantics as in the PostgreSQL specification.
+// In pacticular the methods:
+//  - prepareTxn : https://www.postgresql.org/docs/current/sql-prepare-transaction.html
+//  - commitPreparedTxn : https://www.postgresql.org/docs/current/sql-commit-prepared.html
+//  - rollbackPreparedTxn : https://www.postgresql.org/docs/current/sql-rollback-prepared.html
+
+export interface Database<QueryT, ResultT> {
+
+    beginTxn(): Promise<void>    // initializes a new transaction
+    abortTxn(): Promise<void>    // aborts the current transaction
+
+    runSql(query: QueryT): Promise<ResultT>   // executed a sql statement
+
+    prepareTxn(txnId: string): Promise<void>          // prepared txn for commit under given ID
+    commitPreparedTxn(txnId: string): Promise<void>   // commits a txn prepared under given ID
+    rollbackPreparedTxn(txnId: string): Promise<void> // rolls bacl a txn prepared under given ID
+}
+
+
+export async function runXaDatabaseTransaction<QueryT, ResultT>(
+        ctx: restate.RpcContext,
+        database: Database<QueryT, ResultT>,
+        query: QueryT): Promise<ResultT> {
+
+    // we may need multiple attemps, in case of failures
+    // whenever we resume in a state where the commit decision hasn't been committed
+    // in Restate, we abort whatever may have been started and start over
+    while (true) {
+
+        // step 1: generate and store the ID under which the txn will be stored
+        //         the Id is stored in Restate's log, so that we can use it during
+        //         replay to either commit an accepted transaction, or clean up possibly
+        //         prepared txns that we don't want to commit
+        let weRunTheQuery = false;
```

Comment:
WDYT about renaming this variable to `transactionInProgress`? Makes the if statement further down more clear.